File: erpnext/accounts/report/tds_certificate/tds_certificate.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["TDS Certificate"] = {
	"filters": [
		{
			"fieldname": "branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"options": "Branch",
			"reqd": 1
		},
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1,
			"on_change": function(query_report) {
				var fiscal_year = query_report.get_values().fiscal_year;
				if (!fiscal_year) {
					return;
				}
				frappe.model.with_doc("Fiscal Year", fiscal_year, function(r) {
					var fy = frappe.model.get_doc("Fiscal Year", fiscal_year);
					query_report.filters_by_name.from_date.set_input(fy.year_start_date);
					query_report.filters_by_name.to_date.set_input(fy.year_end_date);
					query_report.trigger_refresh();
				});
			}
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_start_date"),
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_end_date"),
		},
		{
			"fieldname": "vendor_name",
			"label": __("Vendor Name"),
			"fieldtype": "Link",
			"options": "Supplier",
			"on_change": function(query_report) {
				var vendor = query_report.get_values().vendor_name;
				if (!vendor) {
					return;
				}
				frappe.model.with_doc("Supplier", vendor, function(r) {
					var fy = frappe.model.get_doc("Supplier", vendor);
					query_report.filters_by_name.vendor_tpn_no.set_input(fy.vendor_tpn_no);
					query_report.trigger_refresh();
				});
			}
		},
		{
			"fieldname": "vendor_tpn_no",
			"label": __("Vendor TPN Number"),
			"fieldtype": "Data",
			"read_only": 1
		},
	],
}

File: erpnext/accounts/report/tds_certificate/tds_certificate.py
# For license information, please see license.txt
# Frappe Technologiss Pvt. Ltd. and contributors
	
from __future__ import unicode_literals
import frappe

def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data
	
def get_columns():
	return [
		("Invoice Date") + ":Date:120",
		("Gross Amount") + ":Currency:120",
		("tds_rate") + ":Data:80",
		("TDS Amount") + ":Currency:100",
		("Cheque Number") + ":Data:100",
		("Cheque Date") + ":Date:80",
		("Receipt Number") + ":Data:100",
		("Receipt Date") + ":Date:100",
		("Branch") + ":Link/Branch:120"
	]

def get_data(filters):
	query = """SELECT a.posting_date, a.tds_taxable_amount, a.tds_rate, a.tds_amount, b.cheque_number, b.cheque_date, 
		b.receipt_number, b.receipt_date, b.branch 
		FROM `tabPurchase Invoice` AS a, `tabRRCO Receipt Entries` AS b 
		WHERE a.name = b.purchase_invoice 
		AND a.posting_date BETWEEN '{0}' AND '{1}'
		AND a.supplier = '{2}' 
		UNION 
		SELECT  
			d.posting_date,  
			if(d.single_party_multiple_payments = 1, d.taxable_amount, di.taxable_amount) as tds_taxable_amount, d.tds_percent as tds_rate,  
			if(d.single_party_multiple_payments = 1, d.tds_amount, di.tds_amount) as tds_amount, 
			rr.cheque_number,  rr.cheque_date, rr.receipt_number, rr.receipt_date, rr.branch 
		FROM 
			`tabDirect Payment` AS d
		INNER JOIN
			`tabRRCO Receipt Entries` AS rr ON rr.purchase_invoice = d.name 
		LEFT JOIN 
			`tabDirect Payment Item` AS di ON di.parent = d.name 
		WHERE d.posting_date BETWEEN '{0}' AND '{1}'
		AND (
			(d.single_party_multiple_payments = 1 AND d.party = '{2}')
			OR
			(d.single_party_multiple_payments = 0 AND  di.party = '{2}')
		)""".format(filters.from_date, filters.to_date, filters.vendor_name)
	return frappe.db.sql(query)
